src/main/java/org/opensearch/neuralsearch/processor/TextEmbeddingProcessor.java

```diff
@@ -0,0 +1,269 @@
+        log.info("Text embedding result fetched, starting build vector output!");
+        Map<String, Object> vectorOutput = buildVectorOutput(knnMap, vectors, ingestDocument.getSourceAndMetadata());
+        vectorOutput.forEach(ingestDocument::appendFieldValue);
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    private List<String> createInferenceList(Map<String, Object> knnMap) {
+        List<String> texts = new LinkedList<>();
+        knnMap.entrySet().stream().filter(knnMapEntry -> knnMapEntry.getValue() != null).forEach(knnMapEntry -> {
+            Object sourceValue = knnMapEntry.getValue();
+            if (sourceValue instanceof List) {
+                ((List<String>) sourceValue).stream().filter(StringUtils::isNotBlank).forEach(texts::add);
+            } else if (sourceValue instanceof Map) {
+                processMapTypeInput(sourceValue, texts);
+            } else {
+                texts.add(sourceValue.toString());
+            }
+        });
+        return texts;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void processMapTypeInput(Object sourceValue, List<String> texts) {
+        if (sourceValue instanceof Map) {
+            ((Map<String, Object>) sourceValue).forEach((k, v) -> processMapTypeInput(v, texts));
+        } else if (sourceValue instanceof List) {
+            ((List<String>) sourceValue).stream().filter(StringUtils::isNotBlank).forEach(texts::add);
+        } else {
+            if (sourceValue == null || StringUtils.isBlank(sourceValue.toString())) return;
+            texts.add(sourceValue.toString());
+        }
+    }
+
+    @VisibleForTesting
+    Map<String, Object> buildKnnMap(IngestDocument ingestDocument, Map<String, Object> fieldMap) {
+        Map<String, Object> sourceAndMetadataMap = ingestDocument.getSourceAndMetadata();
+        Map<String, Object> knnMap = new LinkedHashMap<>();
+        for (Map.Entry<String, Object> fieldMapEntry : fieldMap.entrySet()) {
+            String originalKey = fieldMapEntry.getKey();
+            Object targetKey = fieldMapEntry.getValue();
+            if (targetKey instanceof Map) {
+                Map<String, Object> treeRes = new LinkedHashMap<>();
+                processConfiguredMapType(originalKey, targetKey, sourceAndMetadataMap, treeRes);
+                knnMap.put(originalKey, treeRes.get(originalKey));
+            } else {
+                knnMap.put(String.valueOf(targetKey), sourceAndMetadataMap.get(originalKey));
+            }
+        }
+        return knnMap;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    private void processConfiguredMapType(
+        String parentKey,
+        Object knnKey,
+        Map<String, Object> sourceAndMetadataMap,
+        Map<String, Object> treeRes
+    ) {
+        if (knnKey == null || sourceAndMetadataMap == null) return;
+        if (knnKey instanceof Map) {
+            Map<String, Object> next = new LinkedHashMap<>();
```

Comment:
why we are using LinkedHashMap instead of HashMap only?

Reply:
For map type user input, we're performing a pre-order traversal on this map to extract all the texts to build the list. And when we build the vector response with the vector list, we need to fill the list value into the map with a correct order. A LinkedHashMap can maintain the order of the keys when iteration on it, but HashMap not.